### PR TITLE
GPflow glossary document

### PR DESCRIPTION
**Pull request by awav**
This glossary should ease reading GPflow's documentation and the code, explaining some non-trivial definitions and abbreviations. For example, we use some uppercase letters like N, L, M, D and others implying that reader familiar with their meaning, of course it is not true. To avoid cumbersome and tedious repetitions in the docs, I created a single definition list markdown file where all well established definitions must take place.

### CODE REVIEW

File: GLOSSARY.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<dl>
```
Comment (st--) [inline GLOSSARY.md]: Would be good to have a header & brief explanation. E.g.:

```
Glossary
========

GPflow does not always follow standard Python naming conventions,
and instead tries to follow the notation in the relevant GP papers.
The following is the convention we aim to follow in the code:
```

File: GLOSSARY.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  <dt>GPR</dt>
+  <dd>gaussian process regression</dd>
```
Comment (st--) [inline GLOSSARY.md]:
```suggestion
  <dd>Gaussian process regression</dd>
```

File: GLOSSARY.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  <dt>SVGP</dt>
+  <dd>stochastic variational inference for Gaussian process models</dd>
+
+  <dt>[] shape constructions</dt>
```
Comment (st--) [inline GLOSSARY.md]:
```suggestion
  <dt>`[...]` shape annotations</dt>
```

File: GLOSSARY.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  <dd>the way of describing tensor shape in docstrings and comments. Example: [..., N, D, D], this is tensor with unknown number of leading dimensions and last two dimensions are equal</dd>
```
Comment (st--) [inline GLOSSARY.md]:
```suggestion
  <dd>The way of describing tensor shapes in docstrings and comments. Example: [..., N, D, D], this is a tensor with an arbitrary number of leading dimensions, and the last two dimensions are equal</dd>
```

File: GLOSSARY.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+  <dt>X</dt>
+  <dd>and its variations like Xnew in GPflow code mean input data (numpy or tensorflow)</dd>
```
Comment (st--) [inline GLOSSARY.md]:
```suggestion
  <dd>(and variations like Xnew) refers to input points; always of rank 2, e.g. shape [N, D], even when D=1</dd>
```
(I think a lot of the entries can describe both numpy arrays or tensorflow tensors so I think it's confusing if this is the only place where it's mentioned explicitly)

File: GLOSSARY.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+  <dt>Y</dt>
+  <dd>throughout is used to denote observed data, potentially with multiple output dimensions</dd>
```
Comment (st--) [inline GLOSSARY.md]:
```suggestion
  <dd>refers to observed output values, potentially with multiple output dimensions; always of rank 2, e.g. shape [N,
 P], even when P=1</dd>
```
Would it be useful to point towards the notebooks here (Coregionalization still has "P=1", the outputs are distinguished by one of the columns in the input points)?

File: GLOSSARY.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  <dd>stands for data or minibatch size in docstrings and shape constructions</dd>
+
+  <dt>P</dt>
+  <dd>stands for the number of output dimensions in docstrings and shape constructions</dd>
+
+  <dt>D</dt>
+  <dd>stands for the number of input dimensions in docstrings and shape constructions</dd>
+
+  <dt>Z</dt>
+  <dd>inducing points tensor</dd>
```
Comment (st--) [inline GLOSSARY.md]:
```suggestion
  <dd>inducing points</dd>
```

File: GLOSSARY.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+  <dt>...</dt>
+  <dd>leading dimensions in docstrings and in the code it is ellipsis.</dd>
```
Comment (st--) [inline GLOSSARY.md]:
```suggestion
  <dd>arbitrary leading dimensions</dd>
```
(this is true both for docstrings and code)

Comment (st--) [reply GLOSSARY.md]: maybe move this right underneath the entry for shapes ?

File: GLOSSARY.md (hunk 3 above)
Comment (awav) [reply GLOSSARY.md]: @st-- , it is slightly misleading because later we say that the ellipsis sign `...` denotes leading dimensions.

Comment (awav) [reply GLOSSARY.md]: @st-- , but I agree that empty brackets don't tell much as well.

File: GLOSSARY.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+  <dt>X</dt>
+  <dd>(and variations like Xnew) refers to input points; always of rank 2, e.g. shape <i>[N, D]</i>, even when <i>D=1</i></dd>
+
+  <dt>Y</dt>
+  <dd>(and variations like Ynew) refers to observed output values, potentially with multiple output dimensions; always of rank 2, e.g. shape <i>[N, P]</i>, even when <i>P=1</i></dd>
+
+  <dt>Z</dt>
+  <dd>refers to inducing points</dd>
+
```
Comment (st--) [inline GLOSSARY.md]:
```suggestion

  <dt>M</dt>
  <dd>stands for the number of inducing features (e.g. length of Z)</dd>

```

File: GLOSSARY.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  <dt>M</dt>
+  <dd>stands for the number of inducing features (e.g. length of Z)</dd>
+
+  <dt>N</dt>
+  <dd>stands for data or minibatch size in docstrings and shape constructions</dd>
```
Comment (awav) [inline GLOSSARY.md]:
```suggestion
  <dd>stands for the number of data or minibatch size in docstrings and shape constructions</dd>
```

File: README.md [hunk 11]
```diff
@@ -6,7 +6,8 @@
 
 [Website](https://gpflow.org) |
 [Documentation](https://gpflow.readthedocs.io/en/master/) |
-[Tips & Tricks](https://github.com/GPflow/GPflow/blob/develop/doc/source/notebooks/tips_and_tricks.ipynb)
+[Glossary](GLOSSARY.md) |
```
Comment (awav) [inline README.md]:
```suggestion
[Glossary](GLOSSARY.md) |```

**Pull request merged**